wgpu-native/src/device.rs

```diff
@@ -210,11 +210,18 @@ impl DestroyedResources<back::Backend> {
             let resource_id = self.mapped.swap_remove(i).value;
             let buf = &buffer_guard[resource_id];
 
-            let usage = match buf.pending_map_operation {
-                Some(BufferMapOperation::Read(..)) => resource::BufferUsageFlags::MAP_READ,
-                Some(BufferMapOperation::Write(..)) => resource::BufferUsageFlags::MAP_WRITE,
-                _ => unreachable!(),
-            };
+            let mut usage = resource::BufferUsageFlags::empty();
+
+            for pending_operation in buf.pending_map_operations.iter() {
+                usage |= match pending_operation {
+                    BufferMapOperation::Read(..) => resource::BufferUsageFlags::MAP_READ,
+                    BufferMapOperation::Write(..) => resource::BufferUsageFlags::MAP_WRITE,
+                };
+                if usage.contains(resource::BufferUsageFlags::MAP_READ | resource::BufferUsageFlags::MAP_WRITE) {
```

Comment:
as the code stands now, wouldn't this condition always be true? we just "|"-ed the usage with either READ or WRITE

Reply:
This just breaks out of the loop early if `usage` has already accumulated both `MAP_READ` and `MAP_WRITE`. 

It's effectively `if usage.contains(MAP_READ) && usage.contains(MAP_WRITE) { break }` . 

The bitwise-or does make it read a little weird.

Reply:
oh snap, you are right! I misread it, sorry :)